### ISSUE TITLE
JSONDecodeError: Expecting value: line 1 column 1 (char 0)

### ISSUE THREAD

**Issue by pruizjun** [GeneralMills/pytrends#122]
Hi,
I'm experiencing a problem with build_payload(). While working on a project, it would arise occasionally and then it became more frequent. Now it occurs every single time I run the program.

Using the example for exposition, I get the following error message:

---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
<ipython-input-2-5fba590191c5> in <module>()
     10 
     11 # Create payload and capture API tokens. Only needed for interest_over_time(), interest_by_region() & related_queries()
---> 12 pytrend.build_payload(kw_list=['pizza', 'bagel'])
     13 
     14 # Interest Over Time

C:\Users\***\Anaconda3\lib\site-packages\pytrends\request.py in build_payload(self, kw_list, cat, timeframe, geo, gprop)
     79         token_payload['req'] = json.dumps(token_payload['req'])
     80         # get tokens
---> 81         self._tokens(token_payload)
     82         return
     83 

C:\Users\***\Anaconda3\lib\site-packages\pytrends\request.py in _tokens(self, token_payload)
     92         # strip off garbage characters that break json parser
     93         widget_json = req.text[4:]
---> 94         widget_dict = json.loads(widget_json)['widgets']
     95         # order of the json matters...
     96         first_region_token = True

C:\Users\***\Anaconda3\lib\json\__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    317             parse_int is None and parse_float is None and
    318             parse_constant is None and object_pairs_hook is None and not kw):
--> 319         return _default_decoder.decode(s)
    320     if cls is None:
    321         cls = JSONDecoder

C:\Users\***\Anaconda3\lib\json\decoder.py in decode(self, s, _w)
    337 
    338         """
--> 339         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    340         end = _w(s, end).end()
    341         if end != len(s):

C:\Users\***\Anaconda3\lib\json\decoder.py in raw_decode(self, s, idx)
    355             obj, end = self.scan_once(s, idx)
    356         except StopIteration as err:
--> 357             raise JSONDecodeError("Expecting value", s, err.value) from None
    358         return obj, end

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

**Comment by xassbit** [comment]
I'm experiencing the same issue. The problem here seems to be that somehow the request is not sent correctly, which returns a 400 error from Google. Since this response is not JSON, it breaks the JSON loader. In my opinion, all non-2** responses should be caught and handled (even if just by throwing a more enlightening exception).

Bellow follows the text of the response from Google:
```html
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 400 (Bad Request)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>400.</b> <ins>That’s an error.</ins>
  <p>Your client has issued a malformed or illegal request.  <ins>That’s all we know.</ins>
```

**Comment by yudhvir** [comment]
@xassbit have you found a fix?
I can't iron out exactly what is causing a problem in the request. According to some other posts in February, a timeframe change seemed to have caused this error, but now there seems to be something new.

**Comment by xassbit** [comment]
The issue seems to be related with the amount of keywords in the payload: I've run some tests and this 400 response is returned when over 5 keywords are added. In our application we started doing requests in batches of 5 keywords and so far we stoped experiencing this issue.

However my previous point still stands: some handler should be added to verify if the content of the response is in fact JSON. I've forked the project and will be working on it - I'll submit a pull when I'm done.

**Comment by 503dannyk** [comment]
I'm currently working on a project utilizing pytrends and found that when I have the following list of keywords it works:
![screen shot 2017-04-09 at 3 44 34 pm](https://cloud.githubusercontent.com/assets/15665979/24841525/9d21c44a-1d3b-11e7-8a5e-e387146ed027.png)

But as soon as I add one additional "+" to the fourth grouping, it throws this exact error. I don't know if this helps, but I would appreciate any suggestions.
![screen shot 2017-04-09 at 3 44 34 pm](https://cloud.githubusercontent.com/assets/15665979/24841548/ee5ff5ac-1d3b-11e7-83b8-201079d10277.png)

**Comment by 503dannyk** [comment]
Update: It looks like any combination of 5 groups of over 15 keywords returns this error. In the example above, if the first five groups contains more than 15 keywords, it returns this error. That is why the addition of "+ Houston Chronicle", created this error. A "+ LA Times" would have the same effect since, it would also have 16 key words.

**Comment by yudhvir** [comment]
Ok, got a handle on the problem.
The JSONDecodeError comes up when you reach a rate limit. This happens when your IP has passed Google's allowed number of search hits per IP for a certain time duration. In my case I was able to hit up 1400+ requests in an hour before the IP got blocked. 

**Note**-Using TOR as a SOCKS proxy won't help here. TOR exit nodes are public information and as such are riddled with reCAPTCHAs replacing the HTML sign in pages. I had a much lower (nearly non-existent) successful response rate while changing TOR circuits. I could only get it to work by first manually passing the sign in page (the reCAPTCHA) and then running pytrends. 

A much better solution is to use a VPN provider. Their proxy IP list is not a source of public information. I used VPN Unlimited and found that none of their IPs were getting blacklisted by Google. Hence, I could do ~7000 searches in a couple of hours (mostly limited by bandwidth itself). 

However, to @xassbit point, there should be a better logging of non-2xx responses and reCAPTCHAs. I recommend closing this issue only once this is done. As far as bypassing the IP limit, a service like [Scrapy](https://github.com/aivarsk/scrapy-proxies) could be used.
